### PR TITLE
GDB component, but it's usable

### PR DESCRIPTION
**Pull request by depau**

### CODE REVIEW

**Comment by depau** [comment]
gdb works (`/x86_64-pc-linux-gnu/x86_64-pc-linux-gnu/binutils-bin/12.1/gdb`) but readline seems a bit borked and it makes the REPLs kinda unusable (both python and gdb), I'll take a look on monday

```
GNU gdb (GDB) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
(gdb) pi
>>> import sys
>>> sys.version
'3.10.6 (main, Aug 25 2022, 16:49:47) [GCC 11.2.0]'
>>> sys.path
['/shared/orchestra/root/share/binutils-data/x86_64-pc-linux-gnu/12.1/gdb/python', '/shared/orchestra/root/lib/python310.zip', '/shared/orchestra/root/lib/python3.10', '/shared/orchestra/root/lib/python3.10/lib-dynload', '/shared/orchestra/root/lib/python3.10/site-packages']
>>>
(gdb) i pretty-printer
global pretty-printers:
  builtin
    mpx_bound128
(gdb)
```

File: .orchestra/config/components/readline.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#@ load("@ytt:data", "data")
+#@ load("@ytt:overlay", "overlay")
+#@ load("/lib/create_component.lib.yml", "single_build_component")
+#@ load("/lib/make.lib.yml", "make")
+
+#@ readline_ver = "8.1.2"
+#@ source_url = "https://ftp.gnu.org/gnu/readline/readline-{}.tar.gz".format(readline_ver)
```
Comment (aleclearmind) [inline .orchestra/config/components/readline.yml]: We use `+`, sad but either convert them all (are fstrings supported?) or stick with old-school string concatenation.

File: .orchestra/config/components/openssl.yml [hunk 2]
```diff
@@ -0,0 +1,53 @@
+#@ load("@ytt:data", "data")
```
Comment (aleclearmind) [inline .orchestra/config/components/openssl.yml:1]: `Co-authored-by: Davide Depau <davide@rev.ng>`?

File: .orchestra/config/components/openssl.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+#@yaml/text-templated-strings
+---
+#@ def openssl_component():
+#@ source_url = "https://github.com/openssl/openssl/archive/refs/tags/OpenSSL_1_1_1q.tar.gz"
+license: LICENSE
+builds:
+  default:
+    build_dependencies:
+      - host-c-toolchain
+      - glibc
```
Comment (aleclearmind) [inline .orchestra/config/components/openssl.yml]: I think we usually keep deps at the end. Push down.

File: .orchestra/config/components/openssl.yml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    configure: |
+      mkdir -p "$BUILD_DIR"
+      extract.sh --into "$BUILD_DIR" (@= source_url @)
+      cd "$BUILD_DIR" && ./Configure \
```
Comment (aleclearmind) [inline .orchestra/config/components/openssl.yml]: Drop `&&`

File: .orchestra/config/components/openssl.yml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        --prefix="$ORCHESTRA_ROOT" \
+        --openssldir="$ORCHESTRA_ROOT/etc/ssl" \
```
Comment (aleclearmind) [inline .orchestra/config/components/openssl.yml]: Maybe we should use `/etc/ssl`. Not sure OpenSSL ships root CAs.
Also it's probably being hardcoded, if we keep it as is, we need to see if there's an environment variable to set to override it.

File: .orchestra/config/components/expat.yml [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+#@ end
+
+#@overlay/match by=overlay.all, expects=1
+#@overlay/match-child-defaults missing_ok=True
+---
+components:
+  expat:
+    license: COPYING
+    default_build: optimized
+    builds: #@ typical_cmake_builds(**build_args())
```
Comment (aleclearmind) [inline .orchestra/config/components/expat.yml:22]: This will create three builds, we don't want this for components that we expect not to change.
However apparently we have a precedent, `libfort`, so I think we can keep it as it is.

File: .orchestra/config/components/bzip2.yml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  
+install: |
+  cd "$BUILD_DIR"
+  
+  # Build
+  (@= make @) -f Makefile-libbz2_so CC="${CC:-gcc} ${CFLAGS:-} ${CPPFLAGS:-} ${LDFLAGS:-}" PREFIX=/
+  (@= make @) bzip2 bzip2recover CC="${CC:-gcc} ${CFLAGS:-} ${CPPFLAGS:-} ${LDFLAGS:-}" PREFIX=/
+
+  # Install
+  PKGDIR="${DESTDIR}${ORCHESTRA_ROOT}"
```
Comment (aleclearmind) [inline .orchestra/config/components/bzip2.yml]: `s/PKG/INSTALL/`

File: .orchestra/config/components/bzip2.yml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  extract.sh --into "$BUILD_DIR" (@= source_url @)
+
+  sed 's/-Winline//g' -i "$BUILD_DIR"/Makefile*
+  
+install: |
+  cd "$BUILD_DIR"
+  
+  # Build
+  (@= make @) -f Makefile-libbz2_so CC="${CC:-gcc} ${CFLAGS:-} ${CPPFLAGS:-} ${LDFLAGS:-}" PREFIX=/
+  (@= make @) bzip2 bzip2recover CC="${CC:-gcc} ${CFLAGS:-} ${CPPFLAGS:-} ${LDFLAGS:-}" PREFIX=/
```
Comment (aleclearmind) [inline .orchestra/config/components/bzip2.yml]: Drop `CC=`. We have our magic magic. We don't need those.

File: .orchestra/config/components/bzip2.yml [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  install -dm755 "$PKGDIR"/{bin,lib64,include,lib64/pkgconfig}
+
+  install -m755 bzip2-shared "$PKGDIR"/bin/bzip2
+  install -m755 bzip2recover bzdiff bzgrep bzmore "$PKGDIR"/bin
+  ln -sf bzip2 "$PKGDIR"/bin/bunzip2
+  ln -sf bzip2 "$PKGDIR"/bin/bzcat
+
+  cp -a libbz2.so* "$PKGDIR"/lib64
+  ln -s libbz2.so.(@= bzip2_ver @) "$PKGDIR"/lib64/libbz2.so
+  ln -s libbz2.so.(@= bzip2_ver @) "$PKGDIR"/lib64/libbz2.so.1 # For compatibility with some other distros
```
Comment (aleclearmind) [inline .orchestra/config/components/bzip2.yml]: Uhm, drop.

File: .orchestra/config/components/libxcrypt.yml [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  popd
+
+  pushd build-libxcrypt-compat
+  ../configure \
+    --prefix="$ORCHESTRA_ROOT" \
+    --disable-static \
+    --enable-hashes=strong,glibc \
+    --enable-obsolete-api=glibc \
+    --disable-failure-tokens
+  popd
```
Comment (aleclearmind) [inline .orchestra/config/components/libxcrypt.yml]: Drop (below too)

File: .orchestra/config/components/python.yml [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        --without-static-libpython
+
+      grep -v HAVE_SYS_AUXV_H pyconfig.h > pyconfig.h.tmp
+      mv pyconfig.h.tmp pyconfig.h
+    install: |
+      cd "$BUILD_DIR"
+      (@= make @)
+      (@= make @) install DESTDIR="$DESTDIR"
+
+      PKGDIR="${DESTDIR}${ORCHESTRA_ROOT}"
```
Comment (aleclearmind) [inline .orchestra/config/components/python.yml]: `PKG` is not an orchestra term

File: .orchestra/config/components/python.yml [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+      ln -s python3               "${PKGDIR}"/bin/python
+      ln -s python3-config        "${PKGDIR}"/bin/python-config
+      ln -s idle3                 "${PKGDIR}"/bin/idle
+      ln -s pydoc3                "${PKGDIR}"/bin/pydoc
```
Comment (aleclearmind) [inline .orchestra/config/components/python.yml]: Drop extra spaces

File: .orchestra/config/components/boost.yml [hunk 13]
```diff
@@ -12,7 +12,7 @@ configure: |
 
   extract.sh --into "$BUILD_DIR" "https://downloads.sourceforge.net/project/boost/boost/1.72.0/boost_1_72_0.tar.bz2"
   patch-if-exists "${ORCHESTRA_DOTDIR}/patches/boost-1.63.0-icl-disable-LessThanComparableConcept.patch" "$BUILD_DIR"
-  cd "$BUILD_DIR" && ./bootstrap.sh --prefix="$ORCHESTRA_ROOT" --with-libraries=test
+  cd "$BUILD_DIR" && ./bootstrap.sh --prefix="$ORCHESTRA_ROOT" --with-libraries=test,regex
```
Comment (aleclearmind) [inline .orchestra/config/components/boost.yml]: State in a comment who needs what.
`test` is needed by revng and revng-c.
`regex`?

File: .orchestra/config/components/source-highlight.yml [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+---
+#@ def source_highlight_args():
+license: COPYING
+configure: |
+  mkdir -p "$BUILD_DIR"
+  extract.sh --into "$BUILD_DIR" (@= source_url @)
+  
+  cd "$BUILD_DIR"
+  
+  sed 's/ throw [(]IOException[)]//g' -i lib/srchilite/fileutil.h lib/srchilite/fileutil.cc
```
Comment (aleclearmind) [inline .orchestra/config/components/source-highlight.yml:16]: What?

File: .orchestra/config/components/toolchain/lib/gdb.lib.yml [hunk 15]
```diff
@@ -42,22 +43,47 @@ configure: |
...
     --enable-nls \
-    --enable-tui=no \
+    --enable-tui \
     --with-mpfr="$ORCHESTRA_ROOT" \
     --without-included-gettext \
     --disable-shared \
     --enable-obsolete \
     --without-stage1-ldflags \
+    --with-python="$ORCHESTRA_ROOT/bin/python" \
+    --with-system-gdbinit="$ORCHESTRA_ROOT/(@= triple @)/etc/gdb/gdbinit" \
```
Comment (aleclearmind) [inline .orchestra/config/components/toolchain/lib/gdb.lib.yml]: These are most likely problematic and non-portable across binary-archives.
I think we should not specify them.

File: .orchestra/config/components/toolchain/lib/gdb.lib.yml [hunk 16]
```diff
@@ -42,22 +43,47 @@ configure: |
...
+  (@= make @) install DESTDIR="$DESTDIR"
+
+  GDBINIT="$DESTDIR$ORCHESTRA_ROOT/(@= triple @)/etc/gdb/gdbinit"
+  mkdir -p "$(dirname "$GDBINIT")"
+
+  cat >"$GDBINIT" <<'EOF'
+  set pagination off
+  EOF
+
+  (@- if triple.startswith("x86_64"): @)
```
Comment (aleclearmind) [inline .orchestra/config/components/toolchain/lib/gdb.lib.yml:95]: There's also `host`. `x86_64` is a cross toolchain.
But I think the `.gdbinit` thing is non-portable and we should just drop it.

File: .orchestra/config/components/toolchain/lib/gdb.lib.yml [hunk 17]
```diff
@@ -42,22 +43,47 @@ configure: |
...
+  (@ end @)
+
+dependencies:
+  - expat
+  - gmp
+  - ncurses
+  - python
+  - readline
+  - source-highlight
+  - toolchain/host/gcc
```
Comment (aleclearmind) [inline .orchestra/config/components/toolchain/lib/gdb.lib.yml]: Please use `host-c-toolchain`. One day we might drop `gcc` as an host toolchain.

File: .orchestra/config/components/source-highlight.yml [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+  cd "$BUILD_DIR"
+  (@= make @)
+  (@= make @) install DESTDIR="$DESTDIR" \
+    bash_completiondir="${DESTDIR}${ORCHESTRA_ROOT}/share/bash-completion/completions"
+
+dependencies:
+  - boost
+
+build_dependencies:
+  - host-c-toolchain
```
Comment (aleclearmind) [inline .orchestra/config/components/source-highlight.yml]: `host-cxx-toolchain`

File: .orchestra/config/components/toolchain/lib/gdb.lib.yml (hunk 17 above)
Comment (depau) [reply .orchestra/config/components/toolchain/lib/gdb.lib.yml]: It doesn't need the compiler (not as a non-build dependency at least), it needs `gcc-libs` and this component provides them.

File: .orchestra/config/components/toolchain/lib/gdb.lib.yml (hunk 16 above)
Comment (depau) [reply .orchestra/config/components/toolchain/lib/gdb.lib.yml:95]: No, `host`'s `triple` variable starts with `x86_64` and I specifically looked for something to specifically recognize all x86_64 targets.

File: .orchestra/config/components/source-highlight.yml (hunk 14 above)
Comment (depau) [reply .orchestra/config/components/source-highlight.yml:16]: `throw` specifiers are deprecated since C++11 and removed in C++17 https://en.cppreference.com/w/cpp/language/except_spec

So either this needs to be done or an old C++ version needs to be forced.

Arch chooses to remove the throw specifiers so that's what I did https://github.com/archlinux/svntogit-packages/blob/packages/source-highlight/trunk/source-highlight-gcc11.patch

File: .orchestra/config/components/openssl.yml (hunk 2 above)
Comment (depau) [reply .orchestra/config/components/openssl.yml:1]: I did minor changes to the file so I set you as author and me as co-author

File: .orchestra/config/components/readline.yml (hunk 1 above)
Comment (depau) [reply .orchestra/config/components/readline.yml]: Starlark does not support f-strings. Using `+` is very sad though

File: .orchestra/config/components/bzip2.yml (hunk 8 above)
Comment (depau) [reply .orchestra/config/components/bzip2.yml]: The bzip2 makefile disregards those variables, though. Does the magic™ handle it?

File: .orchestra/config/components/readline.yml (hunk 1 above)
Comment (aleclearmind) [reply .orchestra/config/components/readline.yml]: Use `+` or fix all other places too to use format strings.
Arguably anonymous `.format` is less readable than concatenation.

File: .orchestra/config/components/openssl.yml (hunk 2 above)
Comment (aleclearmind) [reply .orchestra/config/components/openssl.yml:1]: Ahah, OK.

File: .orchestra/config/components/bzip2.yml (hunk 8 above)
Comment (aleclearmind) [reply .orchestra/config/components/bzip2.yml]: Our magic uses a compiler wrapper that use a set of custom environment variables to bypass all the build systems and have homogeneous compiler flags in all places.
tl;dr: yeah, we handle it.

File: .orchestra/config/components/toolchain/lib/gdb.lib.yml (hunk 17 above)
Comment (aleclearmind) [reply .orchestra/config/components/toolchain/lib/gdb.lib.yml]: Uhm, I wonder why. Can you link the archlinux package you got insipration from?
These are Gentoo's deps:
```
RDEPEND="
        dev-libs/mpfr:0=
        dev-libs/gmp:=
        >=sys-libs/ncurses-5.2-r2:0=
        >=sys-libs/readline-7:0=
        sys-libs/zlib
        elibc_glibc? ( net-libs/libnsl:= )
        lzma? ( app-arch/xz-utils )
        python? ( ${PYTHON_DEPS} )
        guile? ( >=dev-scheme/guile-2.0 )
        xml? ( dev-libs/expat )
        source-highlight? (
                dev-util/source-highlight
        )
        xxhash? (
                dev-libs/xxhash
        )
"
DEPEND="${RDEPEND}"
BDEPEND="
        app-arch/xz-utils
        sys-apps/texinfo
        virtual/yacc
        nls? ( sys-devel/gettext )
        source-highlight? ( virtual/pkgconfig )
        test? ( dev-util/dejagnu )
"
```

File: .orchestra/config/components/openssl.yml (hunk 5 above)
Comment (depau) [reply .orchestra/config/components/openssl.yml]: I think in general it's indeed safest to use the distro/user provided certs anyway

File: .orchestra/config/components/toolchain/lib/gdb.lib.yml (hunk 17 above)
Comment (depau) [reply .orchestra/config/components/toolchain/lib/gdb.lib.yml]: It needs `libstdc++` which is provided by that binary archive.

https://github.com/archlinux/svntogit-packages/blob/packages/gdb/trunk/PKGBUILD#L49

```
$ pacman -Qo /usr/lib/libstdc++.so.6
/usr/lib/libstdc++.so.6 is owned by gcc-libs 12.2.0-1
```

I guess we can statically link to it though.

Comment (depau) [reply .orchestra/config/components/toolchain/lib/gdb.lib.yml]: Ah no it also uses `libgcc_s.so.1`

File: .orchestra/config/components/toolchain/lib/gdb.lib.yml (hunk 15 above)
Comment (depau) [reply .orchestra/config/components/toolchain/lib/gdb.lib.yml]: It actually appears to be portable:

```
$ sudo mount --bind /shared/orchestra/root/ /mnt/

$ cat /mnt/x86_64-pc-linux-gnu/etc/gdb/gdbinit
set pagination off
set disassembly-flavor intel
p "hello world"

$ /mnt/x86_64-pc-linux-gnu/x86_64-pc-linux-gnu/binutils-bin/12.1/gdb -q
$1 = "hello world"
(gdb) pi
>>> import sys
>>> sys.executable
'/mnt/bin/python'
>>> sys.path
['/mnt/share/binutils-data/x86_64-pc-linux-gnu/12.1/gdb/python', '/mnt/lib/python310.zip', '/mnt/lib/python3.10', '/mnt/lib/python3.10/lib-dynload', '/mnt/lib/python3.10/site-packages']
>>>
(gdb) quit
```

Comment (depau) [reply .orchestra/config/components/toolchain/lib/gdb.lib.yml]: Since I specifically need both of these for my use-case, I'd really like to keep them ;)

Comment (aleclearmind) [reply .orchestra/config/components/toolchain/lib/gdb.lib.yml]: Eh no, it's a bug in the orchestra design, can't accept it.
For your use case, just put the stuff in `~/.gdbinit`.

Comment (depau) [reply .orchestra/config/components/toolchain/lib/gdb.lib.yml]: What about python?

File: .orchestra/config/components/openssl.yml (hunk 5 above)
Comment (depau) [reply .orchestra/config/components/openssl.yml]: I changed it to `/etc/ssl` but now a question arises: what if the distro does not use `/etc/ssl`? I was thinking that we should instead provide our own fake config that tries to include all distro configs

File: .orchestra/config/components/toolchain/lib/gdb.lib.yml (hunk 15 above)
Comment (aleclearmind) [reply .orchestra/config/components/toolchain/lib/gdb.lib.yml]: If it spawns a new `python`, it should look it up in the `PATH`.
If it uses `libpython.so` it should already fetch the right one.

Basically, every time we specify a full path in `./configure` for non-build time stuff, it's a bug.

File: .orchestra/config/components/openssl.yml (hunk 5 above)
Comment (aleclearmind) [reply .orchestra/config/components/openssl.yml]: You can check `/etc/ssl` exists before invoking configure and error out if not.

File: .orchestra/config/components/toolchain/lib/gdb.lib.yml (hunk 15 above)
Comment (depau) [reply .orchestra/config/components/toolchain/lib/gdb.lib.yml]: I just read the `configure` source and it appears that the GDB authors thought about this problem already and made it so that if these directories are under `$PREFIX` they are automatically made relative, which is why my example works.

Let me see if there's a way to make it more obvious that this is the case.

File: .orchestra/config/components/openssl.yml (hunk 5 above)
Comment (depau) [reply .orchestra/config/components/openssl.yml]: Yes but this won't prevent the user from installing it from a binary archive

File: .orchestra/config/components/toolchain/lib/gdb.lib.yml (hunk 15 above)
Comment (aleclearmind) [reply .orchestra/config/components/toolchain/lib/gdb.lib.yml]: Ah that's very nice! A comment will do.

File: .orchestra/config/components/openssl.yml (hunk 5 above)
Comment (aleclearmind) [reply .orchestra/config/components/openssl.yml]: Ah eh, you're right. We'd need a post-unpack phase, which we don't have. Leave a comment.
Do you actually know a distro not using `/etc/ssl`?

Comment (depau) [reply .orchestra/config/components/openssl.yml]: NixOS? Guix? Android? (though we'd have many more problems with Android)

I guess though that users of these distros would be used to troubleshooting problems caused by their usage of non-posix-compliant paths.

File: .orchestra/config/components/toolchain/lib/gdb.lib.yml [hunk 19]
```diff
@@ -27,38 +35,69 @@ configure: |
...
+dependencies:
+  - expat
+  - gmp
+  - ncurses
+  - python
+  - readline
+  - source-highlight
 build_dependencies:
+  - host-c-toolchain
+  - host-cxx-toolchain
```
Comment (aleclearmind) [inline .orchestra/config/components/toolchain/lib/gdb.lib.yml]: One implies the other. If `gdb` is not written in C++, don't use `host-cxx-toolchain`. Even if source-highlight is in C++, it'll be its own responsibility to pull in the dependency.

**Comment by depau** [comment]
I pushed the changes that fix the dependency on the glibc 2.32 symbol and, since we're gonna have to rebuild everything anyway, I bumped the other trivial libraries as well.

The issues with readline/ncurses/terminfo are also fixed and all REPLs now appear to be usable.

GDB now pulls all dependencies and the stuff that was being implicitly enabled (since it was available on my system) are being explicitly disabled.

The changes can be reviewed.

File: .orchestra/config/components/boost.yml [hunk 20]
```diff
@@ -5,14 +5,18 @@
 
 #@yaml/text-templated-strings
 ---
+#@ # Boost libraries usages:
+#@ # - test: revng, revng-c
+#@ # - regex: source-highlight
+
```
Comment (aleclearmind) [inline .orchestra/config/components/boost.yml]: Push this right above the libraries list or we'll forget to update it.

**Comment by aleclearmind** [comment]
The xcrypt situation, are we still fetching the system? One can you recap in a comment if something wrong is going on and why?

**Comment by depau** [comment]
> The xcrypt situation, are we still fetching the system? One can you recap in a comment if something wrong is going on and why?

I'm building only the old one `libcrypt.so.1`.

The problem is that Python depends on `-lcrypt` (no reason I'm aware of why it should use ABI v1 and not v2) but I just can't get it to link to `libcrypt.so.2` for some reason.

Nothing else (afaik) depends on lib(x)crypt.

**Comment by aleclearmind** [comment]
OK, by "in a comment" I meant in the code

**Comment by depau** [comment]
> OK, by "in a comment" I meant in the code

Does it work now? 0622158

File: .orchestra/config/components.yml [hunk 21]
```diff
@@ -18,6 +18,7 @@ environment:
   - YARN_RC_FILENAME: $ORCHESTRA_ROOT/share/yarnrc
   - NODE_PATH: $ORCHESTRA_ROOT/node_modules${NODE_PATH:+:${NODE_PATH}}
   - LIBRARY_PATH: /usr/lib/x86_64-linux-gnu
+  - TERMINFO: $ORCHESTRA_ROOT/share/terminfo
```
Comment (depau) [inline .orchestra/config/components.yml:20]: ~~@aleclearmind I was thinking isn't it better if the `ncurses` component provides var this instead? (it's possible to make a component add envs, isn't it?)~~

Nevermind, I assumed that since it's possible to add things to `$PATH` it was also possible to add arbitrary envs.

File: .orchestra/config/components/expat.yml [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+
+---
+#@ def build_args():
+test: false
+source_url: "https://github.com/libexpat/libexpat/releases/download/R_2_4_8/expat-2.4.8.tar.gz"
+use_asan: false
+build_dependencies:
+  - cmake
+  - glibc
+  - host-cxx-toolchain
```
Comment (aleclearmind) [inline .orchestra/config/components/expat.yml]: expat is C.

File: .orchestra/config/components/expat.yml [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+#@ load("/lib/cmake.lib.yml", "typical_cmake_builds")
+
+---
+#@ def build_args():
+test: false
+source_url: "https://github.com/libexpat/libexpat/releases/download/R_2_4_8/expat-2.4.8.tar.gz"
+use_asan: false
+build_dependencies:
+  - cmake
+  - glibc
```
Comment (aleclearmind) [inline .orchestra/config/components/expat.yml]: I think `glibc` is implied by `host-c-toolchain`, you can drop it. One day we might switch to musl `:)`.

File: .orchestra/config/components/python.yml [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+        --libdir="${ORCHESTRA_ROOT}/lib64" \
+        --enable-optimizations \
+        --enable-shared \
+        --enable-ipv6 \
+        --disable-test-modules \
+        --with-ensurepip=install \
+        --with-lto \
+        --with-readline \
+        --with-system-ffi \
+        --with-system-expat \
```
Comment (aleclearmind) [inline .orchestra/config/components/python.yml:29]: Why did you choose to provide expat instead of using the builtin one?

File: .orchestra/config/components/source-highlight.yml [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+#@ def source_highlight_args():
+license: COPYING
+configure: |
+  mkdir -p "$BUILD_DIR"
+  extract.sh --into "$BUILD_DIR" (@= source_url @)
+  
+  cd "$BUILD_DIR"
+  sed 's/ throw [(]IOException[)]//g' -i lib/srchilite/fileutil.h lib/srchilite/fileutil.cc
+  
+  CC=clang CXX=clang++ CXXFLAGS=-stdlib=libc++ ./configure \
```
Comment (aleclearmind) [inline .orchestra/config/components/source-highlight.yml]: Drop `CXXFLAGS=-stdlib=libc++`, our magic compiler wrapper will provide it silently.
```
.orchestra/config/global_options.yml:hard_flags_cxx_clang: -stdlib=libc++
```

File: .orchestra/config/components/source-highlight.yml [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+    --with-boost="$ORCHESTRA_ROOT" \
+    --with-sysroot="$ORCHESTRA_ROOT" \
+    --without-doxygen \
+    --disable-static
+
+install: |
+  cd "$BUILD_DIR"
+  (@= make @)
+  (@= make @) install DESTDIR="$DESTDIR" \
+    bash_completiondir="${DESTDIR}${ORCHESTRA_ROOT}/share/bash-completion/completions"
```
Comment (aleclearmind) [inline .orchestra/config/components/source-highlight.yml]: We don't ship bash and this is prolly getting hardcoded.
Drop. Ibidem above.

File: .orchestra/ci/install-system-dependencies.sh [hunk 27]
```diff
@@ -41,6 +41,7 @@ apt-get -qq install --no-install-recommends --yes \
   flex \
   g++-multilib \
   gawk \
+  gettext \
```
Comment (aleclearmind) [inline .orchestra/ci/install-system-dependencies.sh]: We're going to merge soon a branch where for each dep you need to say why it's needed.
Who needs this?

**Review by aleclearmind — CHANGES_REQUESTED** [review]

File: .orchestra/config/components/python.yml (hunk 24 above)
Comment (depau) [reply .orchestra/config/components/python.yml:29]: Because we're going to use it for gdb anyway.

File: .orchestra/ci/install-system-dependencies.sh (hunk 27 above)
Comment (depau) [reply .orchestra/ci/install-system-dependencies.sh]: glibc wants `msgfmt` to build

**Comment by depau** [comment]
All comments addressed

**Comment by aleclearmind** [comment]
Putting this on hold, waiting for 22.04 CI branch to go in.

**Comment by aleclearmind** [comment]
Branch went in, please rebase.

**Comment by depau** [comment]
Rebased

**Comment by aleclearmind** [comment]
Merged in cd310d5. Thanks!
